### PR TITLE
CAL-23 Fixes for NSILI source for remote queries and product downloads.

### PR DESCRIPTION
**Pull request by troymohl**
#### What does this PR do?
Fixes for the NSILI source on executing remote queries and downloading products.
#### Who is reviewing it?
@bdeining 
@kcwire 
@jaymcnallie 
@jlcsmith 
#### How should this be tested?
JUnit Tests
Integration with NSILI/STANAG-4559 system.
#### Any background context you want to provide?
The dependent PRs need to be merged before this will work.
https://github.com/codice/ddf/pull/635 -> Needs to be in place for the product downloads to work correctly if the source system returns a HTTP Redirect, not required to build.
#### What are the relevant tickets?
CAL-23
DDF-1588 -> Fix product downloads to support redirects, not required to build.
#### Checklist:
- [ ] Documentation Updated
- [X] Update / Add Unit Tests
   -> Unit tests have been updated to reflect the changes in behavior required.
- [ ] Update / Add Integration Tests

### CODE REVIEW

File: catalog/nsili/catalog-nsili-common/src/main/java/org/codice/alliance/nsili/common/NsiliMetacardType.java [hunk 1]
```diff
@@ -35,6 +35,11 @@
     public static final String NSILI_METACARD_TYPE_POSTFIX = "metacard";
 
     /**
+     * The Metacard Type name for NSILI base class
+     */
+    public static final String METACARD_TYPE_NAME = NSILI_METACARD_TYPE_PREFIX+"."+NSILI_METACARD_TYPE_POSTFIX;
```
Comment (bdeining) [inline catalog/nsili/catalog-nsili-common/src/main/java/org/codice/alliance/nsili/common/NsiliMetacardType.java]: Formatting (whitespace)

File: distribution/sdk/sample-nsili-server/src/main/java/org/codice/alliance/nsili/mockserver/server/MockWebService.java [hunk 2]
```diff
@@ -43,7 +43,7 @@ public Response getProductFile() {
     public Response getIORFile() {
         File file = new File(System.getProperty("user.dir") + IOR_RELATIVE_PATH);
         Response.ResponseBuilder response = Response.ok(file);
-        response.header("Content-Disposition", "attachment; filename=ior.txt");
+//        response.header("Content-Disposition", "attachment; filename=ior.txt");
```
Comment (bdeining) [inline distribution/sdk/sample-nsili-server/src/main/java/org/codice/alliance/nsili/mockserver/server/MockWebService.java]: Can we just remove this commented out code?

File: distribution/sdk/sample-nsili-server/src/main/java/org/codice/alliance/nsili/mockserver/client/NsiliClient.java [hunk 3]
```diff
@@ -123,6 +123,8 @@ public void initLibrary(String iorFilePath) {
 
     public void initManagers(String[] managers) throws Exception {
         for (String managerType : managers){
+            System.out.println("Attempting to initialize manager type: "+managerType);
```
Comment (bdeining) [inline distribution/sdk/sample-nsili-server/src/main/java/org/codice/alliance/nsili/mockserver/client/NsiliClient.java]: Formatting

Comment (troymohl) [inline distribution/sdk/sample-nsili-server/src/main/java/org/codice/alliance/nsili/mockserver/client/NsiliClient.java]: Fixed

File: distribution/sdk/sample-nsili-server/src/main/java/org/codice/alliance/nsili/mockserver/server/MockWebService.java (hunk 2 above)
Comment (troymohl) [inline distribution/sdk/sample-nsili-server/src/main/java/org/codice/alliance/nsili/mockserver/server/MockWebService.java]: Yeah. I meant to remove it prior to checkin.

File: catalog/nsili/catalog-nsili-common/src/main/java/org/codice/alliance/nsili/common/NsiliMetacardType.java (hunk 1 above)
Comment (troymohl) [inline catalog/nsili/catalog-nsili-common/src/main/java/org/codice/alliance/nsili/common/NsiliMetacardType.java]: Fixed

File: catalog/nsili/catalog-nsili-common/src/main/java/org/codice/alliance/nsili/common/NsiliAttributeMap.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+import ddf.catalog.data.Metacard;
+
+public class NsiliAttributeMap {
+    private static Map<String, String> nsiliToDdfMap = new HashMap<>();
+    private static Map<String, List<String>> ddfToNsiliMap = new HashMap<>();
+
+    static {
+        //Build up attribute mappings
+
+        nsiliToDdfMap.put("NSIL_APPROVAL.approvedBy", "approvedBy");
```
Comment (bdeining) [inline catalog/nsili/catalog-nsili-common/src/main/java/org/codice/alliance/nsili/common/NsiliAttributeMap.java]: Some of these prefixes and attributes are already defined in NsiliConstants, can we use them instead and add the ones that don't exist into that class?

Comment (troymohl) [inline catalog/nsili/catalog-nsili-common/src/main/java/org/codice/alliance/nsili/common/NsiliAttributeMap.java]: Good catch. That's a good idea. I caught the Metacard side and completely missed the NSIL side.

File: catalog/nsili/catalog-nsili-source/src/main/java/org/codice/alliance/nsili/source/NsiliFilterFactory.java [hunk 5]
```diff
@@ -103,11 +105,32 @@ public String buildPropertyIsLike(String value) {
         // Replace * with %, since * is not a valid wildcard in BQS
         value = value.replaceAll("\\*", "%");
```
Comment (bdeining) [inline catalog/nsili/catalog-nsili-source/src/main/java/org/codice/alliance/nsili/source/NsiliFilterFactory.java]: NsiliFilterDelegate.WILDCARD could be used here

File: catalog/nsili/catalog-nsili-source/src/main/java/org/codice/alliance/nsili/source/NsiliSource.java [hunk 6]
```diff
@@ -855,7 +873,9 @@ private void availabilityChanged(boolean isAvailable) {
      */
     private SortAttribute[] getSortAttributes(SortBy sortBy) {
         if (sortBy == null || sortableAttributes == null) {
-            return new SortAttribute[0];
+            //Default to sorting by Date/Time modified if no sorting provided
+            return new SortAttribute[] {new SortAttribute(NsiliConstants.NSIL_CARD+"."+NsiliConstants.DATE_TIME_MODIFIED,
```
Comment (bdeining) [inline catalog/nsili/catalog-nsili-source/src/main/java/org/codice/alliance/nsili/source/NsiliSource.java]: Formatting

Comment (troymohl) [inline catalog/nsili/catalog-nsili-source/src/main/java/org/codice/alliance/nsili/source/NsiliSource.java]: Fixed

**Comment by bdeining** [comment]
LGTM

File: catalog/nsili/catalog-nsili-source/src/main/java/org/codice/alliance/nsili/source/NsiliFilterFactory.java [hunk 7]
```diff
@@ -101,13 +103,34 @@ public String buildPropertyIsLike(String value) {
...
 
         // Replace * with %, since * is not a valid wildcard in BQS
-        value = value.replaceAll("\\*", "%");
-        value = value.replaceAll("\\?", "%");
+        value = value.replaceAll("\\*", NsiliFilterDelegate.WILDCARD);
+        value = value.replaceAll("\\?", NsiliFilterDelegate.WILDCARD);
+        boolean addPrefixWildcard = false;
+        boolean addPostfixWildcard = false;
+
+        if (!value.startsWith("%")) {
```
Comment (bdeining) [inline catalog/nsili/catalog-nsili-source/src/main/java/org/codice/alliance/nsili/source/NsiliFilterFactory.java]: Could use WILDCARD here as well

File: catalog/nsili/catalog-nsili-source/src/main/java/org/codice/alliance/nsili/source/NsiliFilterFactory.java [hunk 8]
```diff
@@ -101,13 +103,34 @@ public String buildPropertyIsLike(String value) {
...
+            addPrefixWildcard = true;
+        }
+
+        if (!value.endsWith("%")) {
```
Comment (bdeining) [inline catalog/nsili/catalog-nsili-source/src/main/java/org/codice/alliance/nsili/source/NsiliFilterFactory.java]: WILDCARD

File: catalog/nsili/catalog-nsili-common/src/main/java/org/codice/alliance/nsili/common/NsiliMetacardType.java [hunk 9]
```diff
@@ -696,7 +707,6 @@ public AttributeFormat getAttributeFormat() {
     /**
      * NSIL_COMMON Attributes
      * descriptonAbstract -> Metacard.Description
```
Comment (bdeining) [inline catalog/nsili/catalog-nsili-common/src/main/java/org/codice/alliance/nsili/common/NsiliMetacardType.java]: Typo `decripton`?

Comment (troymohl) [inline catalog/nsili/catalog-nsili-common/src/main/java/org/codice/alliance/nsili/common/NsiliMetacardType.java]: Fixed

File: catalog/nsili/catalog-nsili-source/src/main/java/org/codice/alliance/nsili/source/NsiliFilterFactory.java (hunk 7 above)
Comment (troymohl) [inline catalog/nsili/catalog-nsili-source/src/main/java/org/codice/alliance/nsili/source/NsiliFilterFactory.java]: Changed this and the one below to use WILDCARD.

**Comment by shane-voisard** [comment]
Hero build successful of f3383821da87d1de7d956d08836cb18880ef350f and 2cec2d472e1cd579c7c98f830e07e06217dd2a6e . 

added the NSILI app to DDF and then configured the federated source with the following IOR URL:  http://10.101.1.219:9091/nsili/csd/ior

confirmed successful search of the nsili federated source.

**Comment by jaymcnallie** [comment]
:+1:

**Pull request merged**